components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodPortsService.kt

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote
+
+import com.jetbrains.rd.util.URI
+import java.util.Optional
+import java.util.regex.Pattern
+
+class GitpodPortsService {
+    private val hostToClientForwardedPortMap: MutableMap<Int, Int> = mutableMapOf()
```

Comment:
Is not JB already exposes such map to us?

Duplicating and synching content is a recipe for out of sync and race conditions usually. Could we use JB services as a source of the truth instead?

Reply:
I've checked, but unfortunately, they don't.

Here are all 4 methods provided by their PortManager at the moment of writing:
<img width="1170" alt="image" src="https://user-images.githubusercontent.com/418083/182357207-35f83e7b-1213-4670-90b3-605520c43633.png">

And this is what they provide in the PortForwarding package (I've also looked in the sub-packages):
<img width="996" alt="image" src="https://user-images.githubusercontent.com/418083/182358456-ac92ae81-9f7b-4511-a2e7-17dad6d10c50.png">

Note: This is in the snapshot of 2022.2 (which is used in 'io.gitpod.jetbrains.remote.latest' package).
I haven't checked 2022.3 snapshot yet, because EAP JB IDEs on Gitpod.io are still 2022.2.

Reply:
`getForwardedPortsWithLabel` should do?

Reply:
That's what I was expecting too (the `ForwardedPortInfo` to have a `clientPort`), but it doesn't.
The objects from the `getForwardedPortsWithLabel` map have only these properties:
<img width="604" alt="image" src="https://user-images.githubusercontent.com/418083/182800453-7fc6cb3d-8356-478d-beeb-11cd28642a58.png">

Reply:
Could you check with JB folks why is it? For now we can go with your solution. But we should talk with them to refine APIs.